File: bin/send_diff.pl
#!/usr/bin/env perl

=head1 NAME

send_diff.pl - Send email about changes to owner of Netspoc-Web

=head1 COPYRIGHT AND DISCLAIMER

(C) 2020 by Heinz Knutzen     <heinz.knutzen@gmail.com>

https://github.com/hknutzen/Netspoc-Web

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.
You should have received a copy of the GNU General Public License
along with this program. If not, see <http://www.gnu.org/licenses/>.

=cut

use strict;
use warnings;

use utf8;
use Email::Simple;
use Email::Stuffer;
use FindBin;
use lib $FindBin::Bin;
use Load_Config;
use User_Store;
use Template;
use JSON_Cache;
use Policy_Diff;
use open qw(:std :utf8);

# Argument processing
sub usage {
    die "Usage: $0 yyyy-mm-dd yyyy-mm-dd|pxxx\n";
}

my $old_ver = shift @ARGV or usage();
my $new_ver = shift @ARGV or usage();
@ARGV and usage();

my $config = Load_Config::load();
my $path = $config->{netspoc_data};

# Cache holding JSON data.
my $cache = JSON_Cache->new(netspoc_data => $path, max_versions => 8);

# Bestimme aktive owner und zugehörige User (= emails).
# Untersuche, über welche Owner die User informiert werden wollen.
# Das ist im User-Store im Attribut 'send_diff' gespeichert.
my $users_dir = $config->{user_dir};
my $email2owners = $cache->load_json_version($new_ver, 'email');
my %owner2send;

opendir(my $dh, $users_dir) or exit;
for my $email (sort readdir($dh)) {
    $email =~ /[@]/ or next;
    my $wildcard = $email =~ s/^.*@/[all]@/r;
    my %valid;
    if (my $owners = $email2owners->{$wildcard}) {
        @valid{@$owners} = @$owners;
    }
    if (my $owners = $email2owners->{$email}) {
        @valid{@$owners} = @$owners;
    }

    # Bestimme für jeden Benutzer die Owner,
    # über die er bei Änderungen informiert werden möchte.
    my $store = User_Store::load($config, $email) or next;
    my $send_aref = $store->param('send_diff') or next;
    my @send_ok = grep { $valid{$_} } @$send_aref;
    for my $owner (@send_ok) {
        push(@{ $owner2send{$owner} }, $email);
    }

    # Alte, ungültige Owner herausnehmen,
    # da der Benutzer inzwischen weniger Rechte haben könnte
    # oder der Owner nicht mehr existiert.
    # Den Empfänger per Mail über entfernte Owner informieren.
    next if @send_ok == @$send_aref;
    for my $invalid (grep { not $valid{$_} } @$send_aref) {
        my $template
            = -d "$path/current/owner/$invalid"
            ? "$config->{mail_template}/diff_owner_invisible"
            : "$config->{mail_template}/diff_owner_unknown";

        my $text = Template::get($template,
                                 {email => $email, owner => $invalid});
        sendmail($text);
    }
    $store->param('send_diff', \@send_ok)
}
closedir($dh);

my %replace = (
    '+' => '(+)',
    '-' => '(-)',
    '!' => '(!)',
    'service_lists owner' => 'Liste eigener Dienste',
    'service_lists user'  => 'Liste genutzter Dienste',
    'objects'             => 'Objekte',
    'services'            => 'Dienste',
    'users'               => 'Liste der Benutzer (User)',
    );

sub convert {
    my ($in, $level) = @_;
    my $type = ref($in);
    if (not $type) {
        $in = $replace{$in} || $in;
        return(' ' x $level . $in);
    }
    elsif ($type eq 'HASH') {
        return(map { convert($_, $level+1) }
               map { ($_, $in->{$_}) } sort keys %$in);
    }
    elsif ($type eq 'ARRAY') {
        return(map { convert($_, $level+1) } @$in);
    }
}

my %toplevel_sort = (objects => 1, services => 2, );

# Vergleiche Dateien für jeden Owner.
for my $owner (sort keys %owner2send) {

    my $changes = Policy_Diff::compare($cache, $old_ver, $new_ver, $owner)
        or next;
    my $diff = join("\n", map( { convert({ $_ => $changes->{$_} }, -1) }
                               (sort { ($toplevel_sort{$a} || 999)
                                       <=>
                                       ($toplevel_sort{$b} || 999) }
                                keys %$changes)));

    my $template = "$config->{mail_template}/diff";
    for my $email (@{ $owner2send{$owner} }) {
        my $text = Template::get($template,
                                 { email   => $email,
                                   owner   => $owner,
                                   old_ver => $old_ver,
                                   new_ver => $new_ver,
                                   diff    => $diff
                                 });
        sendmail($text);

    }
}

sub sendmail {
    my ($text) = @_;
    my $sendmail = $config->{sendmail_command} or
        die "Missing config for: sendmail_command\n";
    my $from = $config->{noreply_address};
    my $email = Email::Simple->new($text);

    # -t: read recipient address from mail text
    # -f: set sender address
    # -F: do not use sender full name
    open(my $mail, '|-', "$sendmail -t -F '' -f $from")
        or die "Can't open $sendmail: $!";
    print $mail Email::Stuffer->new({ to => $email->header('to'),
                                      subject => $email->header('subject'),
                                      text_body => $email->body(),
                                    })->as_string();
    close $mail or warn "Can't close $sendmail: $!";
}

File: go/pkg/backend/send-diff.go
package backend

import (
	"cmp"
	"fmt"
	"maps"
	"os"
	"slices"
	"strings"
)

func usage() {
	abort("Usage: send-diff yyyy-mm-dd yyyy-mm-dd|pxxx\n")
}

func SendDiff() {
	config := LoadConfig()
	path := config.NetspocData
	s := &state{
		config: config,
		cache:  newCache(config.NetspocData, 8),
	}
	if len(os.Args) != 3 {
		usage()
	}
	oldVer := os.Args[1]
	newVer := os.Args[2]

	userStoreDir := config.UserDir
	emailToOwners := s.loadEmail2Owners()

	userDir, err := os.Open(userStoreDir)
	if err != nil {
		abort("Error opening user directory: %v", err)
	}
	defer userDir.Close()

	users, err := userDir.Readdirnames(0)
	if err != nil {
		abort("Error reading user directory: %v", err)
	}

	ownerToSend := make(map[string][]string)
	for _, email := range users {
		if !strings.Contains(email, "@") {
			continue
		}
		wildcard := "[all]@" + email[strings.Index(email, "@")+1:]
		valid := make(map[string]bool)

		if owners, exists := emailToOwners[wildcard]; exists {
			for _, owner := range owners {
				valid[owner] = true
			}
		}
		if owners, exists := emailToOwners[email]; exists {
			for _, owner := range owners {
				valid[owner] = true
			}
		}
		userFile := fmt.Sprintf("%s/%s", userStoreDir, email)
		store, err := GetUserStore(userFile)
		if err != nil {
			fmt.Printf("Error loading user store for %s: %v\n", email, err)
			continue
		}
		if len(store.SendDiff) < 1 {
			continue
		}
		sendOk := []string{}
		for _, send := range store.SendDiff {
			if valid[send] {
				//fmt.Printf("Sending diff to %s for owner %s\n", email, send)
				sendOk = append(sendOk, send)
			}
		}

		for _, owner := range sendOk {
			ownerToSend[owner] = append(ownerToSend[owner], email)
		}

		// Alte, ungültige Owner herausnehmen,
		// da der Benutzer inzwischen weniger Rechte haben könnte
		// oder der Owner nicht mehr existiert.
		// Den Empfänger per Mail über entfernte Owner informieren.
		if len(sendOk) == len(store.SendDiff) {
			continue
		}
		for _, removed := range store.SendDiff {
			if !valid[removed] {
				fmt.Printf("Not sending diff to %s for removed owner %s\n", email, removed)

				var template string
				if _, err := os.Stat(fmt.Sprintf("%s/current/owner/%s", path, removed)); err == nil {
					template = fmt.Sprintf("%s/diff_owner_invisible", config.MailTemplate)
				} else {
					template = fmt.Sprintf("%s/diff_owner_unknown", config.MailTemplate)
				}
				text, err := s.getTemplateContent(template, map[string]string{
					"Email": email,
					"Owner": removed,
				})
				if err != nil {
					fmt.Printf("Failed to get email template for %s: %v\n", email, err)
					continue
				}
				err = s.sendEmail(text)
				if err != nil {
					fmt.Printf("Failed to send email to %s: %v\n", email, err)
					continue
				}
			}
		}
		store.SendDiff = sendOk
		err = store.WriteToFile(userFile)
		if err != nil {
			fmt.Printf("Failed to write user store for %s: %v\n", email, err)
			continue
		}
	}

	for owner := range ownerToSend {
		changes, err := s.compare(oldVer, newVer, owner)
		if err != nil || changes == nil {
			continue
		}
		diff := ""
		// Sort changes by toplevel keys.
		for _, key := range []string{"objects", "services", "service_lists owner", "service_lists user", "users"} {
			block := make(map[string]interface{})
			block[key] = changes[key]
			if block[key] == nil {
				continue
			}
			for _, line := range Convert(block, 0) {
				diff += line + "\n"
			}
			diff += "\n"
		}
		template := fmt.Sprintf("%s/diff", config.MailTemplate)
		for _, email := range ownerToSend[owner] {
			text, err := s.getTemplateContent(template, map[string]string{
				"Email":  email,
				"Owner":  owner,
				"OldVer": oldVer,
				"NewVer": newVer,
				"Diff":   diff,
			})
			if err != nil {
				fmt.Printf("Failed to get email template for %s: %v\n", email, err)
				continue
			}
			fmt.Fprintln(os.Stderr, "\n", text)
			/*
				err = s.sendEmail(text)
				if err != nil {
					fmt.Printf("Failed to send email to %s: %v\n", email, err)
					continue
				}
			*/
		}
	}
}

var replace = map[string]string{
	"+":                   "(+)",
	"-":                   "(-)",
	"!":                   "(!)",
	"service_lists owner": "Liste eigener Dienste",
	"service_lists user":  "Liste genutzter Dienste",
	"objects":             "Objekte",
	"services":            "Dienste",
	"users":               "Liste der Benutzer (User)",
}

// Convert should convert the changes coming from the func compare.
// The keys of the nested data structure should be changed in-place
// to more human-readable strings according to map "replace".
// The level parameter indicates the
// nesting level (0: top-level, 1: inside objects/services, etc.).
// Each replaced key should be prefixed with indentation spaces
// according to the nesting level.
// The result is a flat list of strings representing the changes.
// Each replaced key is on a new separate line.
func Convert(input interface{}, level int) []string {
	var result []string
	indent := strings.Repeat("  ", level)

	switch v := input.(type) {
	case map[string]interface{}:
		keys := slices.SortedFunc(maps.Keys(v), func(a, b string) int {
			return cmp.Compare(strings.ToLower(a), strings.ToLower(b))
		})
		for _, key := range keys {
			value := v[key]
			replacedKey := replace[key]
			if replacedKey == "" {
				replacedKey = key
			}
			result = append(result, fmt.Sprintf("%s%s", indent, replacedKey))
			switch nestedValue := value.(type) {
			case map[string]interface{}:
				result = append(result, Convert(nestedValue, level+1)...)
			case map[string][]string:
				for key, values := range nestedValue {
					replacedKey := replace[key]
					if replacedKey == "" {
						replacedKey = key
					}
					result = append(result, fmt.Sprintf("%s%s", indent, replacedKey))
					for _, val := range values {
						replacedValue := replace[val]
						if replacedValue == "" {
							replacedValue = val
						}
						result = append(result, fmt.Sprintf("%s%s", indent+"  ", replacedValue))
					}
				}
			case []interface{}:
				result = append(result, Convert(nestedValue, level+1)...)
			case []string:
				for _, val := range nestedValue {
					replacedValue := replace[val]
					if replacedValue == "" {
						replacedValue = val
					}
					result = append(result, fmt.Sprintf("%s%s", indent+"  ", replacedValue))
				}
			case string:
				replacedValue := replace[nestedValue]
				if replacedValue == "" {
					replacedValue = nestedValue
				}
				result = append(result, fmt.Sprintf("%s%s", indent+"  ", replacedValue))
			case int:
				result = append(result, fmt.Sprintf("%s%d", indent+"  ", nestedValue))
			case map[int]interface{}:
				result = append(result, Convert(nestedValue, level+1)...)
			default:
				fmt.Fprintf(os.Stderr, "Unhandled nested value type %T in Convert\n", nestedValue)
				result = append(result, fmt.Sprintf("%s%v", indent+"  ", nestedValue))
			}
		}
	case []interface{}:
		for _, item := range v {
			result = append(result, Convert(item, level)...)
		}
	case string:
		replacedValue := replace[v]
		if replacedValue == "" {
			replacedValue = v
		}
		result = append(result, fmt.Sprintf("%s%s", indent, replacedValue))
	}

	return result
}
